### PR TITLE
Initialize-from custom  checkpoints 

### PR DESCRIPTION
**Pull request by mahon94**
### Proposed change(s)

Initialize from custom checkpoints

Usage: 
```
behaviors:
  BigWallJump:
    Init_path: BigWallJump-654098.pt #specify a checkpoint_name in the behavior directory
    trainer_type: ppo
  MediumWallJump:
    Init_path: results/previous_run/MediumWallJump/MediumWallJump-654098.pt # full path is also supported
    trainer_type: ppo
SmallWallJump: # if none specified, will initialize from most recent checkpoint
    trainer_type: ppo
checkpoint_settings:
  initialize_from: previous_run
```
TODO: Extra tests, Documentation, changeLog
### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/trainer/trainer_factory.py [hunk 1]
```diff
@@ -100,8 +104,6 @@ def _initialize_trainer(
         :return:
         """
         trainer_artifact_path = os.path.join(output_path, brain_name)
-        if init_path is not None:
-            trainer_settings.init_path = os.path.join(init_path, brain_name)
```
Comment (mahon94) [inline ml-agents/mlagents/trainers/trainer/trainer_factory.py:104]: Moved this logic to learn.py

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 2]
```diff
@@ -75,16 +76,20 @@ def initialize_or_load(self, policy: Optional[TorchPolicy] = None) -> None:
...
     def _load_model(
         self,
         load_path: str,
         policy: Optional[TorchPolicy] = None,
         reset_global_steps: bool = False,
     ) -> None:
-        model_path = os.path.join(load_path, "checkpoint.pt")
-        saved_state_dict = torch.load(model_path)
+        print("loading: ", load_path)
+        saved_state_dict = torch.load(load_path)
```
Comment (mahon94) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:92]: load_path is always the full path to the .pt file. It originally comes from trainer_settings.init_path which is assigned in learn.py

File: ml-agents/mlagents/trainers/learn.py [hunk 3]
```diff
@@ -49,6 +50,23 @@ def parse_command_line(argv: Optional[List[str]] = None) -> RunOptions:
     return RunOptions.from_argparse(args)
 
 
+def _Get_checkpoint_name(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:53]:
```suggestion
def _get_checkpoint_name(
```
snake_case_for_python_functions

File: ml-agents/mlagents/trainers/learn.py [hunk 4]
```diff
@@ -49,6 +50,23 @@ def parse_command_line(argv: Optional[List[str]] = None) -> RunOptions:
...
+    behavior_name: str, checkpoint_list: Optional[Dict[str, str]]
+) -> str:
+    if checkpoint_list and checkpoint_list.get(behavior_name):
+        return checkpoint_list[behavior_name]
+    else:
+        return "checkpoint.pt"
+
+
+def _Validate_init_full_path(init_file: str) -> None:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:62]:
```suggestion
def _validate_init_full_path(init_file: str) -> None:
```

File: ml-agents/mlagents/trainers/learn.py [hunk 5]
```diff
@@ -49,6 +50,23 @@ def parse_command_line(argv: Optional[List[str]] = None) -> RunOptions:
     return RunOptions.from_argparse(args)
 
 
+def _Get_checkpoint_name(
+    behavior_name: str, checkpoint_list: Optional[Dict[str, str]]
+) -> str:
+    if checkpoint_list and checkpoint_list.get(behavior_name):
+        return checkpoint_list[behavior_name]
+    else:
+        return "checkpoint.pt"
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:59]: This is a constant, I think it is hard coded somewhere else in the `torch_model_saver.py` script. Please use a constant so we can change it at only one place.

File: ml-agents/mlagents/trainers/learn.py [hunk 6]
```diff
@@ -49,6 +50,23 @@ def parse_command_line(argv: Optional[List[str]] = None) -> RunOptions:
     return RunOptions.from_argparse(args)
 
 
+def _Get_checkpoint_name(
+    behavior_name: str, checkpoint_list: Optional[Dict[str, str]]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:54]: argument is called checkpoint_list but it is a mapping from str to str. Please rename argument

File: ml-agents/mlagents/trainers/learn.py [hunk 7]
```diff
@@ -49,6 +50,23 @@ def parse_command_line(argv: Optional[List[str]] = None) -> RunOptions:
...
+) -> str:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/learn.py:55]: add some docstrings on these methods (even though they are private)

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 8]
```diff
@@ -25,6 +25,7 @@ def __init__(
         super().__init__()
         self.model_path = model_path
         self.initialize_path = trainer_settings.init_path
+        print("TMS:", self.initialize_path)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:28]:
```suggestion
```

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 9]
```diff
@@ -75,16 +76,20 @@ def initialize_or_load(self, policy: Optional[TorchPolicy] = None) -> None:
             )
         elif self.load:
             logger.info(f"Resuming from {self.model_path}.")
-            self._load_model(self.model_path, policy, reset_global_steps=reset_steps)
+            self._load_model(
+                os.path.join(self.model_path, "checkpoint.pt"),
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:80]: "checkpoint.pt" should be a constant, not a hard coded string.

File: ml-agents/mlagents/trainers/model_saver/torch_model_saver.py [hunk 10]
```diff
@@ -75,16 +76,20 @@ def initialize_or_load(self, policy: Optional[TorchPolicy] = None) -> None:
...
 
     def _load_model(
         self,
         load_path: str,
         policy: Optional[TorchPolicy] = None,
         reset_global_steps: bool = False,
     ) -> None:
-        model_path = os.path.join(load_path, "checkpoint.pt")
-        saved_state_dict = torch.load(model_path)
+        print("loading: ", load_path)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/model_saver/torch_model_saver.py:91]:
```suggestion
```

File: ml-agents/mlagents/trainers/trainer/trainer_factory.py (hunk 1 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/trainer/trainer_factory.py:104]: Have you tried using init_path and see if it does what we expect out of it?
Reading a bit of the code in `torch_model_saver.py` it looks like setting init_path in the trainer settings initializes the policy from a checkpoint. Is that correct?

**Review by vincentpierre — CHANGES_REQUESTED** [review]
I made some comments. 
I would like some clarification on what init_path in the trainer parameters do. It seems a bit redundant with the checkpoint settings.

File: ml-agents/mlagents/trainers/trainer/trainer_factory.py (hunk 1 above)
Comment (mahon94) [reply ml-agents/mlagents/trainers/trainer/trainer_factory.py:104]: Correct. Previously `trainer_settings.init_path="result_dir/run_id/brain_name` and we added the checkpoint name at the end in `torch_model_saver.py`
This PR sets the full path `trainer_settings.init_path="result_dir/run_id/brain_name/checkpoint_name.pt` in [learn.py:102](https://github.com/Unity-Technologies/ml-agents/blob/b0c2361031af87551dc7ecf5595a43020651e09e/ml-agents/mlagents/trainers/learn.py#L102)

**Review by vincentpierre — COMMENTED** [review]
I approve of the code.
Got one question before you merge. If the user fills both init_path and the checkpoint_list. Which one gets priority.
(Please add documentation and changelog before merging)

File: ml-agents/mlagents/trainers/tests/test_trainer_util.py [hunk 11]
```diff
@@ -137,3 +141,47 @@ def test_existing_directories(tmp_path):
...
+        big.write("content")
+        med = init_path.mkdir("MediumWallJump").join("checkpoint.pt")
+        med.write("content")
+        small = init_path.mkdir("SmallWallJump").join("checkpoint.pt")
+        small.write("content")
+
+        setup_init_path(run_options.behaviors, init_path)
+        assert run_options.behaviors["BigWallJump"].init_path == big
+        assert run_options.behaviors["MediumWallJump"].init_path == med
+        assert run_options.behaviors["SmallWallJump"].init_path == small
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/test_trainer_util.py:181]: probably should delete these files and folders after the assert call. I would also give a different name than "results" to the directory. Something like `test_setup_init_path_results`

Comment (mahon94) [reply ml-agents/mlagents/trainers/tests/test_trainer_util.py:181]: [pytest](https://docs.pytest.org/en/6.2.x/tmpdir.html#the-default-base-temporary-directory) deletes the stale tmp directories automatically. I'll update the directory name.

Comment (mahon94) [reply ml-agents/mlagents/trainers/tests/test_trainer_util.py:181]: done

**Review by vincentpierre — APPROVED** [review]

**Pull request merged**